---
title: "Homework3"
author: "Aditya Lahiri"
date: "March 20, 2018"
output:
  word_document: default
---

# Question 1
```{r, include=FALSE}
## Use seqinr to obtain the Genbank sequence for PSEN1.
library(seqinr)
choosebank("genbank")
PSEN1 <- query(listname = "PSEN1", query="SP=homo sapiens AND K=PSEN1")
```
## Part a
```{r}
psen1_seqs <- getSequence(PSEN1)
psen1_names <- getName(PSEN1)
```
The number of sequences are:
```{r}
length(psen1_seqs)
```
Their names are as follows:
```{r}
print(psen1_names)
```
Their lengths are as follows:
```{r}
print(unlist(lapply(psen1_seqs,length)))
```

## Part B
The fasta file entry is as follows:
```{r}
write.fasta(psen1_seqs[1], psen1_names[1], file.out = "MyPSEN1.fasta")
```

## Part C
% Content of GC in the third sequence is =
```{r}
sum(table(psen1_seqs[[3]])[2:3]) / length(psen1_seqs[[3]])*100
```


# Question 2
```{r}
library(Biostrings)
seq_1 <- "GTAG"
seq_2 <- "GAG"

score_mat <- nucleotideSubstitutionMatrix(match = 1, mismatch = -1, baseOnly = TRUE)
gap_open <- -2
gap_extend <- -2

```


## Part a



## Part b
The optimal allignment and score are as follows:
```{r}
alignment <- pairwiseAlignment(seq_1, seq_2, substitutionMatrix = score_mat, gapOpening =
                              gap_open, gapExtension = gap_extend,type ="global",scoreOnly=FALSE)             
alignment
```



# Question 3

## Part A
```{r}
seq_1 <- "ASEDLTI"
seq_2 <- "AEEDFGI"

data(PAM30)
score_mat <- "PAM30"

gap_open <- 0
gap_extend <- -2

global_alignment <- pairwiseAlignment(seq_1, seq_2, substitutionMatrix = score_mat,
                                      gapOpening = gap_open, gapExtension = gap_extend, type = "global", scoreOnly = FALSE)
local_alignment <- pairwiseAlignment(seq_1, seq_2, substitutionMatrix = score_mat,
                                     gapOpening = gap_open, gapExtension = gap_extend, type = "local", scoreOnly = FALSE)

```

The Global allignment is as follows:
```{r}
global_alignment

```

The Local allignment is as follows:
```{r}
local_alignment
```

## Part B



## Part C
```{r}
gen_random_seqs <- function(sq, B) {
  ## Break input sequence into vector of characters.
  sq <- strsplit(sq, "")[[1]]
  n <- length(sq)
  ## Compute sample proportions of each observed letter.
  sq_tbl <- table(sq)
  sq_letters <- names(sq_tbl)
  n_letters <- length(sq_letters)
  pp <- numeric()
  for(i in 1:length(sq_letters))
    pp[i] <- sq_tbl[i] / n
  ## Generate B random sequences by sampling with replacement using a multinomial model.
  sqs <- numeric(B)
  for(i in 1:B)
    sqs[i] <- paste(sample(sq_letters, n, rep = TRUE, prob = pp), collapse = "")
  return(sqs)
}

## Function for computing a p-value for the alignment of two sequences. The sequence in
## the first argument should be shorter than or equal in length to that in the second
## argument.
align_p_val <- function(seq_1, seq_2, scoringMat, gapOpen, gapExtend, alignType, B) {
  ## Compute alignment score for original pair of sequences.
  tt_0 <- pairwiseAlignment(seq_1, seq_2, substitutionMatrix = scoringMat,
                            gapOpening = gapOpen, gapExtension = gapExtend, type = "local", scoreOnly = TRUE)
  ## Generate B random versions of first sequence.
  random_seqs <- gen_random_seqs(seq_1, B)
  ## For each random sequence, re-run the alignment and store the resulting score.
  tt_B <- numeric(B)
  for(i in 1:B)
    tt_B[i] <- pairwiseAlignment(random_seqs[i], seq_2, substitutionMatrix = scoringMat,
                                 gapOpening = gapOpen, gapExtension = gapExtend, type = alignType, scoreOnly = TRUE)
  ## Compute a p-value as the proportion of randomization-based alignment scores that are
  ## equal to or greater than our original observed score.
  p_val <- mean(tt_B >= tt_0)
  return(list("tt_0" = tt_0, "tt_B" = tt_B, "p_val" = p_val))
  
}
set.seed(101)
out <- align_p_val(seq_1, seq_2, score_mat, gap_open, gap_extend, "local", 1000)
```

## i
```{r}
dd <- density(out$tt_B)
plot(dd, main = "Randomization-based scores")
#pos_x <- max(which(dd$x >=19))
m <- length(dd$x)
delta <- dd$x - 19
ii <- match(min(delta[delta >= 0]), delta)
polygon(c(dd$x[ii], dd$x[ii:m], dd$x[m]), c(0, dd$y[ii:m], 0), col = "red")

```
## ii
The resulting p value is:
```{r}
out$p_val
```


# Question 4

## Part A

Genes that share an arbitrary threshold level of similarity determined by alignment of matching bases are termed homologous. Homologs have common  origins but may or may not have common activity. [1]
https://msu.edu/~jhjacksn/Reports/similarity.htm

## Part B

```{r}
# download dna sequences  HS -> homo sapiens MM-> MUS MUSCULUS
GENE_HS <- query(listname = "H1F0", query="SP=homo sapiens AND K=H1F0")  
GENE_MM <- query(listname = "H1F0", query="SP=mus musculus AND K=H1F0")

# Get the sequences
HS_H1F0_seqs <- getSequence(GENE_HS) # human sequence
MM_H1F0_seqs <- getSequence(GENE_MM) # mouse sequence

# Sequences
HS_Seq <- HS_H1F0_seqs[1]
MM_Seq <- MM_H1F0_seqs[1]


```

## i.
The % difference in basses is =
```{r}

 Miss_Match <- 100*length(which(HS_Seq[[1]]!=MM_Seq[[1]]))/length(MM_Seq[[1]]) # % Difference in Bases
Miss_Match
```

## ii.
The dot plot is as follows:
```{r}
dotPlot(HS_Seq[[1]][1:585], MM_Seq[[1]][1:585], col=c("white", "red"), xlab = "HUMAN", ylab = "MOUSE",main="DOT PLOT OF MOUSE VS HUMAN MAX GENE")

```
## iii.

```{r}
HS_Seq<-toupper(HS_Seq[[1]])
MM_Seq<-toupper(MM_Seq[[1]])
HS_Seq<-paste(HS_Seq,collapse="")
MM_Seq<-paste(MM_Seq,collapse="")
out <- align_p_val(HS_Seq, MM_Seq, score_mat, gap_open, gap_extend,"global", B = 1000)
```
The Pvalue is 
```{r}
out$p_val
```
## Part C
```{r,include=FALSE}
library(muscle)
GENE_BT <- query(listname = "H1F0", query="SP=bos taurus AND K=H1F0")
BT_H1F0_seqs <- getSequence(GENE_BT) # cow sequence
BT_Seq <- BT_H1F0_seqs[1]
BT_Seq<-toupper(BT_Seq[[1]])
BT_Seq<-paste(BT_Seq,collapse="")
require(Biostrings)
String_Object <-  DNAStringSet(c(BT_Seq,HS_Seq,MM_Seq))
String_align <- muscle(String_Object)
detail(String_align)
```

The first 50 aligned bases are :
```{r}
String_align@unmasked[[1]][1:50]
```

# Question 5
Filter the NAs as needed
```{r}
data_csv = read.csv("qc_spike.csv", header = TRUE)
pos<- vector()
count <-1
for (iter in 1:dim(data_csv)[1]){
  if (sum(is.na(data_csv[iter,8:11])) >2 ){
    pos[count] <- iter
      count = count +1
  }
  else if (sum(is.na(data_csv[iter,12:15])) >2 ) {
     pos[count] <- iter
       count = count +1
  }
  else if (sum(is.na(data_csv[iter,16:19])) >2 ) {
     pos[count] <- iter
       count = count +1
  }

} 

data_csv_filtered <- data_csv[-pos,]
```

## Part A.
The number of peptides left are :
```{r}
dim(data_csv_filtered)[1]
```
The number of QC peptides are:
```{r}
length(which(data_csv_filtered[,7]==1))
```
The number of Salmonella peptides are:
```{r}
length(which(data_csv_filtered[,7]==0))
```


## Part B.
side-by-side boxplots comparing the 12 samples (so, 12 boxplots side-by-side),
for the QC peptide
```{r}
data_mat<-as.matrix(data_csv_filtered[,8:19])
pos_qc <- which(data_csv_filtered[,7]==1)
data_mat_qc<-data_mat[pos_qc,]

boxplot(data_mat_qc,las=2,at =c(1,2,3,4, 6,7,8,9, 11,12,13,14), 
        col=c('red','sienna','palevioletred1','royalblue2','red','sienna','palevioletred1', 
        'royalblue2','red','sienna','palevioletred1','royalblue2'),main="Boxplot of QC  Data")
        

#boxplot.matrix(data_mat)
```

## Part C
side-by-side boxplots comparing the 12 samples, now for the Salmonella peptides
```{r}
pos_salmonella <- which(data_csv_filtered[,7]==0)
data_mat_salmonella<-data_mat[pos_salmonella,]
boxplot(data_mat_salmonella,las=2,at =c(1,2,3,4, 6,7,8,9, 11,12,13,14), 
        col=c('red','sienna','palevioletred1','royalblue2','red','sienna','palevioletred1', 
        'royalblue2','red','sienna','palevioletred1','royalblue2'),main="Boxplot of Salmonella Data")
        
```


## Part D

```{r}

P_values <- vector() # Empty vector
for (iter in 1: dim(data_mat_salmonella)[1]){ # go row by row
  t_stats <-t.test(data_mat_salmonella[iter,c(1,2,3,4)],data_mat_salmonella[iter,c(9,10,11,12)],var.equal = TRUE) # compute t statistic for each row of  high v low concentration
  P_values[iter]<- t_stats$p.value # Store the P-Values 
}
 
```
## i.
histogram of the resulting p-values.
```{r}
hist(P_values) # Plot histogram of P-Values
```

## ii.
proportion of the Salmonella peptides had p-values < 0.05
```{r}
sum(P_values<0.05)
sum(P_values<0.05)/length(P_values)
```

Expected proportion of p-values < 0.05 = 

```{r}
print(sessionInfo())
```
